FAQ for data download troubleshooting (#417)
AlexsLemonade/OpenScPCA-analysis: docs/troubleshooting-faq/faq.md

# Frequently asked questions

### Why didn't the sample/project I specified when running the [data download script](../getting-started/accessing-resources/getting-access-to-data.md#using-the-download-data-script) download?

First, we recommend using the `--dryrun` flag when running the `download-data.py` script to check which files _would_ be downloaded.
This will confirm that there is nothing wrong with your internet connection and that you are properly [logged into your AWS profile](../technical-setup/environment-setup/configure-aws-cli.md#logging-in-to-a-new-session).

If running the script with `--dryrun` states that _only_ the `DATA_USAGE.md` file is being downloaded, this means the data files you are attempting to download do not exist.
There are two main reasons why this might occur:

> 1. Not all projects are available in `AnnData` format.

Reviewer: ```suggestion
1. Not all samples are available in `AnnData` format.
```

If you are attempting to download `AnnData` format, please [consult the ScPCA documentation](https://scpca.readthedocs.io/en/stable/faq.html#which-samples-can-i-download-as-anndata-objects) to learn more about which types of projects do not have `AnnData` files.
1. The sample or project ID(s) that you specified during download do(es) not exist.

Therefore, you may wish to list some of the files in the data release S3 bucket to confirm the sample/project files you are trying to download actually exist.

Data files in each release are organized on S3 as:
```{ .console .no-copy title="Release file structure"}
{Release}
    ├── {Project ID}
    │   └── {Sample ID}
    │       └── {Library files}
    ├── bulk_metadata.tsv (if applicable)
    ├── bulk_quant.tsv (if applicable)
    └── single_cell_metadata.tsv
```

> To list all available files, you'll need to use the AWS CLI [in a terminal](../software-platforms/general-tools/using-the-terminal.md) and run `aws s3 ls` to list the contents of a given S3 bucket.
> Don't forget to [log into your AWS `openscpca` profile first](../technical-setup/environment-setup/configure-aws-cli.md#logging-in-to-a-new-session)!
> 
> 
> 1. First, find the names of all the data releases, which are named based on their release date in [ISO 8601 format](https://en.wikipedia.org/wiki/ISO_8601).
> You can use the [`download-data.py`](../getting-started/accessing-resources/getting-access-to-data.md#download-data-structure) script with the `--list-releases` flag to do this:

Reviewer 2: ```suggestion

1. First, find the names of all the data releases, which are named based on their release date in [ISO 8601 format](https://en.wikipedia.org/wiki/ISO_8601).
You can use the [`download-data.py`](../getting-started/accessing-resources/getting-access-to-data.md#download-data-structure) script with the `--list-releases` flag to do this. 
Don't forget to [log into your AWS `openscpca` profile first](../technical-setup/environment-setup/configure-aws-cli.md#logging-in-to-a-new-session)!
```
It seemed odd to talk about `aws s3 ls` when you don't show an example with it until after listing the releases.

    ```bash
    # Ensure you are in the top-level of the repository
    # cd path/to/OpenScPCA-analysis

    # List all releases in the data release bucket
    # If you're _not_ working on Lightsail for Research, specify your AWS profile name with `--profile openscpca`
    ./download-data.py --list-releases --profile openscpca
    ```

1. Identify the most recent date from the output.
In the example output below, the most recent release is `2024-05-01`.
    ```{ .console .no-copy title="Output from listing all releases"}
    Available release dates:
    2024-03-11
    2024-04-06
    2024-05-01
    ```

> 1. All data is stored on S3 in `s3://openscpca-data-release/{release name}/`.

Reviewer 2: ```suggestion
1. All data is stored on S3 in `s3://openscpca-data-release/{release name}/`.
To list all available files, you'll need to use the AWS CLI [in a terminal](../software-platforms/general-tools/using-the-terminal.md) and run `aws s3 ls` to list the contents of a given S3 bucket.
```

Therefore, you can list files in a given release as, for example,
    ```bash
    # List all releases in the 2024-05-01 release
    # If you're _not_ working on Lightsail for Research, specify your AWS profile name with `--profile openscpca`
    aws s3 ls s3://openscpca-data-release/2024-05-01/ --profile openscpca
    ```

    This has the (abbreviated) output:
    ```{ .console .no-copy title="Output from listing all projects in the 2024-05-01 release"}
    PRE SCPCP000001/
    PRE SCPCP000002/
    PRE SCPCP000003/
    PRE SCPCP000004/
    ```

>     Note that the `PRE` prefix appears for directories that are inside the overall data release bucket, `s3://openscpca-data-release`.

Reviewer: Missed opportunity to be maximally confusing... (note this is _not_ as suggestion)
```
    Note that the `PRE` prefix appears for prefixes that are inside the overall data release bucket, `s3://openscpca-data-release`.
```

Author: I heard you like prefixes

    !!! tip "Don't forget the trailing slash!"

>         The trailing slash at the end of `s3://openscpca-data-release/2024-05-01/` is necessary for contents to be listed.

Reviewer 2: Just noting that you can also use `--recursive` but that will show all files in the directory, which we probably don't want them to do for entire releases.

        If you omit the slash and just run `aws s3 ls s3://openscpca-data-release/2024-05-01 --profile openscpca`, the output will only list the directory itself, and not its contents.

> 2. From there, you can continue to listing the next level of nested files and directories (prefixes).

Reviewer 2: ```suggestion
1. From there, you can continue to list the next level of nested files and directories (prefixes).
```

# Frequently asked questions

### Why didn't the sample/project I specified when running the [data download script](../getting-started/accessing-resources/getting-access-to-data.md#using-the-download-data-script) download?

> First, we recommend using the `--dryrun` flag when running the script to check which files _would_ be downloaded, to confirm that there is nothing wrong with your internet connection and that you are properly [logged into your AWS profile](../technical-setup/environment-setup/configure-aws-cli.md#logging-in-to-a-new-session).

Reviewer 2: ```suggestion
First, we recommend using the `--dryrun` flag when running the script to check which files _would_ be downloaded. 
This will confirm that there is nothing wrong with your internet connection and that you are properly [logged into your AWS profile](../technical-setup/environment-setup/configure-aws-cli.md#logging-in-to-a-new-session).
```

If you specify a sample or project ID that is not present in the OpenScPCA data release, then nothing will download, but the script will not explicitly fail or warn you about this.

Therefore, you may wish to list all files in the data release S3 bucket to confirm the sample/project files you are trying to download exist.
To do this, you'll need to use the AWS CLI [in a terminal](../software-platforms/general-tools/using-the-terminal.md) and run `aws s3 ls` to list the contents of a given S3 bucket.

All data release files are stored in the S3 bucket `s3://openscpca-data-release/`, which you can recursively list the contents of as follows:

1. First, find the names of all the data release buckets, which are named by release date in [ISO 8601 format](https://en.wikipedia.org/wiki/ISO_8601):
    ```bash
    # List all releases in the overall release bucket
    # If you're _not_ working on Lightsail for Research, specify your AWS profile name with `--profile openscpca`

>     aws s3 ls s3://openscpca-data-release/ --profile openscpca

Reviewer: There is a flag for this in the script! It gives a much more user-friendly output:
```
❯ ./download-data.py --list-releases
Available release dates:
2024-03-11
2024-04-06
2024-05-01```

    ```

    !!! tip "Don't forget the trailing slash!"
        The trailing slash at the end of `s3://openscpca-data-release/` is necessary for contents to be listed.

>         If you omit the slash and just run `aws s3 ls s3://openscpca-data-release --profile openscpca`, the output will only list the bucket itself, and not its contents..

Reviewer 2: This isn't true... 
```
allyhawkins@allys-work-mac ~ % aws s3 ls s3://openscpca-data-release --profile openscpca
                           PRE 2024-03-11/
                           PRE 2024-04-06/
                           PRE 2024-05-01/
```

Reviewer: The trailing slash is needed if you want to look at anything below the main bucket:
```
❯ aws s3 ls s3://openscpca-data-release/2024-05-01
                           PRE 2024-05-01/
```
vs.
```
❯ aws s3 ls s3://openscpca-data-release/2024-05-01/
                           PRE SCPCP000001/
                           PRE SCPCP000002/
                           PRE SCPCP000003/
                           PRE SCPCP000004/
                           PRE SCPCP000005/
                           PRE SCPCP000006/
                           PRE SCPCP000007/
                           PRE SCPCP000008/
                           PRE SCPCP000009/
                           PRE SCPCP000010/
                           PRE SCPCP000011/
                           PRE SCPCP000012/
                           PRE SCPCP000013/
                           PRE SCPCP000014/
                           PRE SCPCP000015/
                           PRE SCPCP000016/
                           PRE SCPCP000020/
                           PRE SCPCP000021/
2024-04-30 09:37:14       1691 DATA_USAGE.md
```

    # If you're _not_ working on Lightsail for Research, specify your AWS profile name with `--profile openscpca`
    ./download-data.py --list-releases --profile openscpca
    ```

1. Identify the most recent date from the output.
In the example output below, the most recent release is `2024-05-01`.
    ```{ .console .no-copy title="Output from listing all releases"}
    Available release dates:
    2024-03-11
    2024-04-06
    2024-05-01
    ```

1. You can list files in a given release as, for example,
    ```bash
    # List all releases in the 2024-05-01 release
    # If you're _not_ working on Lightsail for Research, specify your AWS profile name with `--profile openscpca`
    aws s3 ls s3://openscpca-data-release/2024-05-01/ --profile openscpca
    ```

    This has the (abbreviated) output:
    ```{ .console .no-copy title="Output from listing all projects in the 2024-05-01 release"}
    PRE SCPCP000001/
    PRE SCPCP000002/
    PRE SCPCP000003/
    PRE SCPCP000004/
    ```

    !!! tip "Don't forget the trailing slash!"
        The trailing slash at the end of `s3://openscpca-data-release/2024-05-01/` is necessary for contents to be listed.

>         If you omit the slash and just run `aws s3 ls s3://openscpca-data-release/2024-05-01/ --profile openscpca`, the output will only list the directory itself, and not its contents.

Reviewer: ```suggestion
        If you omit the slash and just run `aws s3 ls s3://openscpca-data-release/2024-05-01 --profile openscpca`, the output will only list the directory itself, and not its contents.
```
